src/lib/logical_query_plan/jit_aware_lqp_translator.cpp

```diff
@@ -0,0 +1,273 @@
+#include "jit_aware_lqp_translator.hpp"
+
+#include <boost/range/combine.hpp>
+
+#include <queue>
+#include <unordered_set>
+
+#include "constant_mappings.hpp"
+#include "operators/jit_operator/operators/jit_filter.hpp"
+#include "operators/jit_operator/operators/jit_read_tuple.hpp"
+#include "operators/jit_operator/operators/jit_write_tuple.hpp"
+#include "projection_node.hpp"
+#include "storage/storage_manager.hpp"
+#include "stored_table_node.hpp"
+
+namespace opossum {
+
+std::shared_ptr<AbstractOperator> JitAwareLQPTranslator::translate_node(
+    const std::shared_ptr<AbstractLQPNode>& node) const {
+  uint32_t num_jittable_nodes{0};
+  std::unordered_set<std::shared_ptr<AbstractLQPNode>> input_nodes;
+
+  // Traverse query tree until a non-jitable nodes is found in each branch
+  _breadth_first_search(node, [&](auto& current_node) {
+    if (_node_is_jitable(current_node)) {
+      ++num_jittable_nodes;
+      return true;
+    } else {
+      input_nodes.insert(current_node);
+      return false;
+    }
+  });
+
+  // It does not make sense to create a JitOperatorWrapper for fewer than 2 LQP nodes,
+  // but we may want a better heuristic here
+  if (num_jittable_nodes < 2 || input_nodes.size() != 1) {
```

Comment:
This looks like you cannot deal with `OR`?

Reply:
You mean the `input_nodes.size() != 1` part? `OR`s are represented by `UnionNode`s, but both inputs to the union eventually refer to the same node (i.e., a diamond pattern).